docs/adrs/0012-exposure-events-and-coenrolling-features.md

```diff
@@ -0,0 +1,142 @@
+
+## Pros and Cons of the Options
+
+### Option 1: Record exposure events only when we're sure of which experiment is involved
+
+In practice, this might be only for when exactly one experiment is using the `messaging` feature.
+
+The message sending experiment owner should continue to use the `message-under-experiment` key.
+
+Fewer exposure events will be emitted than should be.
+
+* Good, because this requires little or no client side implementation
+* Good, because this requires little or no data-analysis implementation.
+* Good, because it involves no change to the feature JSON— the user interface for the message sender.
+* Very Bad, because data-analysis result become much less useful, and needs to be aware what other experiments are active.
+
+This is the current implementation.
+
+### Option 2: Record the exposure event with the message key
+
+The Glean event schema for exposure events would have an additional `part-id` extra, or even another type of exposure event created.
+
+The message sending experiment owner should stop using the `message-under-experiment` key.
+
+An exposure event is emitted for _every_ message that is displayed, even if the message came bundled with the app.
+
+An extra manual step in data analysis is needed to get the message keys from the experiment, and then to match up the exposure events to the experiment.
+
+* Bad, because it increases the number of false positive exposure events; increasing bandwidth and compute costs
+* Bad, because it involves a change to the feature JSON, and the user to change their behavior.
+* Bad, because it requires changes to the exposure event schema, which could affect every experiment.
+* Bad, because it requires additional steps to the data-analysis, some of which will be impossible to automate.
+
+### Option 3: Annotate the message with the experiment slug in the feature JSON
+
+Instead of the user telling the feature which _message_ is under experiment, the user annotates the message with the _experiment_ slug.
+
+i.e. remove the `message-under-experiment` property, and replace it with a new optional `experiment` message property.
+
+```json
+{
+    "messages": {
+        "my-first-message": {
+            "experiment": "{experiment}"
+            …
+        }
+    }
+}
+```
+
+When the experiment slug is known, e.g. at enrollment, the feature config is checked for strings containing `{experiment}`, and the slug used to replace it: i.e. the user should not need to type the actual slug.
+
+When the message is displayed, the `experiment` property is the experiment slug; this can be used to look up the branch, and an exposure event then recorded.
+
+* Good, because implementation is relatively simple, and restricted to the client side.
+* Good, because the correct number of exposure events are recorded.
+* Good, because no extra steps needed in post experiment data-analysis.
+* Bad, because it involves a change to the feature JSON, and the user to change their behavior.
+
```

Comment:
Is there an Option 4, which is basically option 3 but we automate annotating the message with the slug so that the user doesn't need to do it each time they create an experiment? I'm worried about the manual effort on the user's part, particularly for something like surveys where they'll copy and relaunch surveys periodically and it would be pretty easy for them to forget to update the slug.

Reply:
May need to talk to Jared on https://mozilla-hub.atlassian.net/browse/EXP-3602 - which reads like co-enrollment should support other features.  There’s too many options already - with sticky and first run and now multi-feature and rollouts.  I don’t think we’re ready to expose this new highly complex pattern beyond messaging.

The change to allow co-enrollment should be the Messaging feature only.  We should not change onboarding to support co-enrollment - or expose this to other features.   This is a common pattern for messages only.  It’s an exception and one I think we want to avoid for others in the foreseeable future.  There is the one possible case for attribution based onboarding - but that is the exception and may be put on hold depending on [these discussions](https://docs.google.com/document/d/13ZFY7Dhprl5LE7tscdl1x9rAu8IPDJ3MN7mIQJraZBQ/edit#heading=h.5ipiksf3te2q).

Reply:
Agree that option 3 looks good, with the caveat of what Daniel commented.  The training and trying to capture human errors is getting overly complex.  The chance for easily done errors is too high - and it would invalidate the experiment learnings if the information isn’t added.

Reply:
The user adds to the message.

```json
"experiment": "{experiment}"
```

I've proposed that the SDK (or experimenter!) does the replacement of the `{experiment}` with the actual slug, so users don't need to manage slugs themselves, and can add it to their own templates.

Reply:
But users still need to remember to add that string template to every message? That's certainly better than adding the exact slug, but is there a way for the SDK or experimenter to do the whole process for the user? If we're going to replace the template with the slug, can we just automate the whole thing?

Reply:
I agree. I don't think `experiment` should be optional